Source/WTF/wtf/SinglyOwned.h

```diff
@@ -0,0 +1,87 @@
+/*
+ *  Copyright (C) 2025 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/RefCountedWithInlineWeakPtr.h>
+#include <wtf/SinglyOwnedPtr.h>
+
+namespace WTF {
+
+// Use CheckedPtr for strong references and InlineWeakPtr for weak references.
+// When a CheckedPtr outlives a SinglyOwnedPtr:
+//     * In debug builds, the CheckedPtr ASSERTs
+//     * In release builds, the CheckedPtr keeps the pointed-to object alive.
+template<typename T>
+class SinglyOwned : private RefCountedWithInlineWeakPtr<T> {
+    friend class RefCountedWithInlineWeakPtr<T>;
+    template<typename U> friend SinglyOwnedPtr<U> adoptSinglyOwned(U*);
+public:
+    // SinglyOwned interface
+    void disown() const
+    {
+        this->refCountDebugger().willDelete();
+        this->deref();
```

Comment:
I think we should be able to detect that this class takes T as an argument and eventually call delete static_cast<T*>(this) via this function in a superclass. That's probably better than hard-coding this class name.